### ISSUE TITLE
`kokkosp_end_deep_copy` needs some more arguments

### ISSUE THREAD

**Issue by khuck** [kokkos/kokkos#5363]
**Describe the bug**
I found a problem in the hpx-kokkos integration (https://github.com/STEllAR-GROUP/hpx-kokkos) that likely will need some fixes from the Kokkos profiling API while using the HPX back end. In the HPX implementation of deep copy (https://github.com/STEllAR-GROUP/hpx-kokkos/blob/master/src/hpx/kokkos/deep_copy.hpp) they create a future to perform an asynchronous deep copy. That's all fine, but the Kokkos profiling interface doesn't provide enough information to profiling tools. Tools implement a callback at the beginning of the deep copy with:
```
void kokkosp_begin_deep_copy(
    SpaceHandle dst_handle, const char* dst_name, const void* dst_ptr,
    SpaceHandle src_handle, const char* src_name, const void* src_ptr,
    uint64_t size) {...}
```
However, the end of the deep copy is:
```
void kokkosp_end_deep_copy() {...}
```
which means that tools don't know which deep copy timer to stop. The assumption was to stop the deep copy on the current posix thread, but because HPX created a future, yielded, and later resumed on another posix thread when the copy was finished, that assumption is broken. Tools would need Kokkos to provide the source/destination handles or addresses to provide me a way to uniquely track the copy event. Either that, or the begin & end events need some kind of unique identifier.

Here's what it looks like in the debugger with the APEX performance tool enabled (notice the deep copy starts on thread 8 and finishes on thread 9):
```
khuck@khuck-VirtualBox:~/build/hpx-kokkos$ export KOKKOS_PROFILE_LIBRARY=/home/khuck/install/hpx/git/lib/libhpx_apex.so
khuck@khuck-VirtualBox:~/build/hpx-kokkos$ gdb --args tests/parallel_algorithms_test 
GNU gdb (Ubuntu 12.0.90-0ubuntu1) 12.0.90
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from tests/parallel_algorithms_test...
(gdb) b kokkosp_begin_deep_copy
Function "kokkosp_begin_deep_copy" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (kokkosp_begin_deep_copy) pending.
(gdb) b kokkosp_end_deep_copy
Function "kokkosp_end_deep_copy" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (kokkosp_end_deep_copy) pending.
(gdb) run
Starting program: /home/khuck/build/hpx-kokkos/tests/parallel_algorithms_test 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff4497640 (LWP 2240)]
stable-f7ec709c62-HEAD
Built on: 13:34:42 Aug 18 2022 (Debug)
C++ Language Standard version : 201703
[New Thread 0x7ffff37ff640 (LWP 2241)]
[New Thread 0x7ffff2ffe640 (LWP 2242)]
[New Thread 0x7ffff1fff640 (LWP 2243)]
[New Thread 0x7ffff15fe640 (LWP 2244)]
[New Thread 0x7ffff03ff640 (LWP 2245)]
[New Thread 0x7fffefbfe640 (LWP 2246)]
[New Thread 0x7fffeedff640 (LWP 2247)]
[New Thread 0x7fffedfff640 (LWP 2248)]
GCC Compiler version : 11.2.0testing executor with execution space "HPX"
[New Thread 0x7fffecbfe640 (LWP 2249)]
[Switching to Thread 0x7fffefbfe640 (LWP 2246)]

Thread 8 "parallel_algori" hit Breakpoint 1, kokkosp_begin_deep_copy (dst_handle=..., dst_name=0x7fffecdfead0 "for_each_data", dst_ptr=0x7fffee02c040, src_handle=..., src_name=0x7fffee0006a0 "for_each_data_host", src_ptr=0x7fffee02bec0, size=172) at /home/khuck/build/hpx/_deps/apex-src/src/apex/apex_kokkos.cpp:324
324	    uint64_t size) {
(gdb) c
Continuing.
[Switching to Thread 0x7fffeedff640 (LWP 2247)]

Thread 9 "parallel_algori" hit Breakpoint 2, kokkosp_end_deep_copy () at /home/khuck/build/hpx/_deps/apex-src/src/apex/apex_kokkos.cpp:344
344	void kokkosp_end_deep_copy() {
(gdb) 
```

**Please also include the following items to support reproducing the bug**
1. compilers (with versions) - any gcc is fine
2. Kokkos release or commit used (i.e., the sha1 number) - current master
3. platform and backend - HPX on any linux
4. cmake configure command 
```
cmake \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_CXX_STANDARD=14 \
-DCMAKE_INSTALL_PREFIX=${HOME}/install/kokkos/git \
-DKokkos_ARCH_HSW=ON \
-DKokkos_ENABLE_HPX=ON \
-DKokkos_ENABLE_HPX_ASYNC_DISPATCH=ON \
-DKokkos_ENABLE_HWLOC=ON \
-DKokkos_ENABLE_TESTS=OFF \
-DKokkos_HWLOC_DIR=$HOME/install/hwloc \
$HOME/src/kokkos
```

5. output from cmake command

6. code needed to reproduce the bug - a working HPX build configured with a simple config (assuming hwloc and lib boost are installed): 
```
cmake \
-D CMAKE_BUILD_TYPE=RelWithDebInfo \
-D CMAKE_INSTALL_PREFIX:PATH=$HOME/install/hpx/git  \
-D HPX_WITH_MALLOC=system  \
-D HWLOC_ROOT=/usr \
-D BOOST_ROOT=/usr \
-D HPX_WITH_NETWORKING=OFF  \
-D HPX_WITH_PARCELPORT_MPI=OFF  \
-D HPX_WITH_EXAMPLES=OFF  \
-D HPX_WITH_TESTS=OFF  \
-D HPX_WITH_APEX=TRUE  \
-D HPX_WITH_APEX_TAG=develop \
-D HPX_WITH_FETCH_ASIO=ON \
$HOME/src/hpx
```
and hpx-kokkos test program tests/parallel_algorithms.cpp, configured with:
```
cmake $HOME/src/hpx-kokkos \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCMAKE_CXX_COMPILER=`which g++` \
    -DHPX_DIR=$HPX_ROOT \
    -DKokkos_DIR=$Kokkos_ROOT \
    -DHPX_KOKKOS_ENABLE_TESTS=ON \
    -DCMAKE_INSTALL_PREFIX=$HOME/install/hpx-kokkos/git
```
7. command line needed to reproduce the bug - `tests/parallel_algorithms` with any Kokkos performance tool that measures deep copies - HPX comes with APEX (and is configured above)
8. please also attach the `KokkosCore_config.h` header file (generated during the build);

**Any additional info**
Please provide any additional context about the issue here.
This is also going to be a problem with `kokkosp_push_profile_region`/`kokkosp_pop_profile_region` - the pop call doesn't have any uniquely identifying information. This would likely only be the case if a user called these functions while using the HPX back end, but other asynchronous back ends could also be a problem. I do realize that the push/pop regions imply using them on the same posix thread.

**Comment by msimberg** [comment]
Thanks @khuck for bringing this up. We've definitely not been active users of the Kokkos profiling hooks with hpx-kokkos so this is undertested.

I think this is orthogonal to hpx-kokkos creating futures for the completion of asynchronous kernels and should be reproducible with vanilla Kokkos+HPX and any parallel pattern as well unless I'm misunderstanding something.

When the HPX backend fences after a parallel pattern (including deep copy) it will likely end up yielding the current HPX thread and when the thread is resumed it may end up running on another worker thread. Currently the only way to guarantee that it resumes on the same thread is to use the static scheduler (or with the bound threads, but that's not very usable currently). Generally, however, we don't want to force HPX threads to resume on the same worker thread where it was suspended. We attempt to because it potentially uses caches better, but we don't make the guarantee. I assume that this would never come up in unit tests in Kokkos since kernels are launched, and fenced, on the main thread and not on an HPX thread.

Like @khuck suggested keeping track of an id is one possibility (an RAII class would make this easier). I don't really see many other options because we'd rather not force an HPX thread to suspend and resume on the same thread.

Do all profiling tools fence internally?

**Comment by dalg24** [comment]
> Do all profiling tools fence internally?

No

**Comment by khuck** [comment]
No, fencing for profiling tools is now optional. By default APEX does not fence, but if we want to use the Kokkos launch to measure the kernel without explicitly measuring the kernel back end task (with HPX timers, CUPTI, Rocprofiler, OMPT, etc.), then we ask Kokkos to fence. See https://github.com/UO-OACISS/apex/blob/41bc390edad478e908911e94db1a9eb18318aef5/src/apex/apex_kokkos.cpp#L155-L161 for example. If the tool doesn't fence, the tool is simply measuring the kernel launch time (and capturing the task dependency graph).

**Comment by msimberg** [comment]
Ok, thanks. I suppose the fencing question was actually a bit off topic.

As far as I can tell, this is in the end simply a feature request to extend the Kokkos profiling interface. At least at the moment I can't think of a (reasonable) way in which this can be fixed in the HPX backend or HPX itself. Do you have ideas that I might not be thinking about? If someone is using the Kokkos profiling tools with HPX one can ask them to use `--hpx:queuing=static`.

@khuck what is the failure mode for this? Segfault? Messed up output files?

**Comment by khuck** [comment]
Correct, this is a request to extend the profiling interface. That way, I can use a process-level map to map the deep copy completion event to the begin event.

The only other option I can think of is that the hpx-kokkos library performs the deep copy synchronously (without a future at all). My fix for right now is to ignore deep copy events when APEX is part of HPX, because I will have to assume that it will just corrupt the profile data.

The result in APEX is not a crash, but that the profile data will be incomplete. A timer will be started on a thread that may never get completed. APEX could potentially give a warning. The way I am handling deep copies currently is with a thread-local timer stack, and that won't work in this case.

**Comment by msimberg** [comment]
> Correct, this is a request to extend the profiling interface. That way, I can use a process-level map to map the deep copy completion event to the begin event.

That sounds good to me at least. As I wrote above I expected this to be a problem for any kernel, but I see now that all the `endParallelX` callbacks have a `kernelID` and `deallocateData` takes the pointer to deallocate. `endDeepCopy` takes nothing, and I guess this is the reason this is only a problem for `deep_copy`?

**Comment by khuck** [comment]
> I guess this is the reason this is only a problem for `deep_copy`

correct... Although, I did mention in the original description: 

> This is also going to be a problem with kokkosp_push_profile_region/kokkosp_pop_profile_region - the pop call doesn't have any uniquely identifying information. This would likely only be the case if a user called these functions while using the HPX back end, but other asynchronous back ends could also be a problem. I do realize that the push/pop regions imply using them on the same posix thread.

If the push/pop profile region is intended to be process-wide, then it would need some special treatment by APEX. If the regions are intended to be thread-specific, then it could potentially be a problem. It's not clear to me what the intention is for this pair of functions...obviously by using push/pop nomenclature, the intent is to build a timer/context stack - but what is the scope: The process? The thread? The current task/function? The entire program/simulation?

**Comment by crtrott** [comment]
I think it makes most sense if its the current thread.
